filepathfilter/filepathfilter.go

```diff
@@ -98,257 +90,72 @@ func (f *Filter) HasPrefix(prefix string) bool {
-		regpattern := fmt.Sprintf("^%s$", strings.Replace(pattern, "\\*\\*", ".*", -1))
-		return &doubleWildcardPattern{
-			rawPattern: cleanpattern,
-			wildcardRE: regexp.MustCompile(regpattern),
-		}
+	if !matched && len(f.include) != 0 {
+		return false
 	}
 
-	if hasPathSep && strings.HasPrefix(cleanpattern, sep) {
-		rel := cleanpattern[1:len(cleanpattern)]
-		prefix := rel
-		if strings.HasSuffix(rel, sep) {
-			rel = rel[0 : len(rel)-1]
-		} else {
-			prefix += sep
-		}
-
-		return &pathPrefixPattern{
-			rawPattern: cleanpattern,
-			relative:   rel,
-			prefix:     prefix,
+	for _, ex := range f.exclude {
+		if ex.HasPrefix(prefix) {
+			return false
 		}
 	}
 
-	return &pathPattern{
-		rawPattern: cleanpattern,
-		prefix:     cleanpattern + sep,
-		suffix:     sep + cleanpattern,
-		inner:      sep + cleanpattern + sep,
-	}
-}
-
-func convertToPatterns(rawpatterns []string) []Pattern {
-	patterns := make([]Pattern, len(rawpatterns))
-	for i, raw := range rawpatterns {
-		patterns[i] = NewPattern(raw)
-	}
-	return patterns
-}
-
-type pathPrefixPattern struct {
-	rawPattern string
-	relative   string
-	prefix     string
-}
-
-// Match is a revised version of filepath.Match which makes it behave more
-// like gitignore
-func (p *pathPrefixPattern) Match(name string) bool {
-	if name == p.relative || strings.HasPrefix(name, p.prefix) {
-		return true
-	}
-	matched, _ := filepath.Match(p.rawPattern, name)
-	return matched
-}
-
-func (p *pathPrefixPattern) HasPrefix(name string) bool {
-	return strings.HasPrefix(p.relative, name)
-}
-
-// String returns a string representation of the underlying pattern for which
-// this *pathPrefixPattern is matching.
-func (p *pathPrefixPattern) String() string {
-	return p.rawPattern
-}
-
-type pathPattern struct {
-	rawPattern string
-	prefix     string
-	suffix     string
-	inner      string
-}
-
-// Match is a revised version of filepath.Match which makes it behave more
-// like gitignore
-func (p *pathPattern) Match(name string) bool {
-	if strings.HasPrefix(name, p.prefix) || strings.HasSuffix(name, p.suffix) || strings.Contains(name, p.inner) {
-		return true
-	}
-	matched, _ := filepath.Match(p.rawPattern, name)
-	return matched
-}
-
-func (p *pathPattern) HasPrefix(name string) bool {
-	return strings.HasPrefix(p.prefix, name)
-}
-
-// String returns a string representation of the underlying pattern for which
-// this *pathPattern is matching.
-func (p *pathPattern) String() string {
-	return p.rawPattern
-}
-
-type simpleExtPattern struct {
-	ext string
-}
-
-func (p *simpleExtPattern) Match(name string) bool {
-	return strings.HasSuffix(name, p.ext)
-}
-
-func (p *simpleExtPattern) HasPrefix(name string) bool {
 	return true
 }
 
-// String returns a string representation of the underlying pattern for which
-// this *simpleExtPattern is matching.
-func (p *simpleExtPattern) String() string {
-	return fmt.Sprintf("*%s", p.ext)
+type wm struct {
+	w    *wildmatch.Wildmatch
+	dirs bool
```

Comment:
Would this be use case for [embedded fields](https://golang.org/ref/spec#Struct_types)?

Reply:
I think that my argument is in the majority motivated by personal preference, but I avoid embedded fields as is possible, since they generate a source of confusion for me. Though verbose, I tend to prefer the explicit nature of `w.w` over `w.<some-inherited-field>`.

Reply:
I agree as I tapped into the embedded field confusion trap before. I wonder if it makes sense to rename`type wm` with `type wm_with_dirs` or something. That's long and ugly, though.
🚲-shedding 😉